modules/solid_mechanics/src/actions/CommonSolidMechanicsAction.C

```diff
@@ -29,8 +31,98 @@ CommonSolidMechanicsAction::CommonSolidMechanicsAction(const InputParameters & p
+    {
+      const auto v = getParam<std::vector<VariableName>>("displacements");
+      if (v.size() > 0)
+        displacement_variables.insert(displacement_variables.end(), v.begin(), v.end());
+    }
+    bool add_variables_block_restricted = true;
+    std::set<SubdomainName> add_variables_blocks;
+    bool scaling_set = isParamSetByUser("scaling");
+    Real scaling_value = (scaling_set) ? getParam<Real>("scaling") : 1.0;
+
+    // check all sub-actions
+    for (const auto & action : actions)
+    {
+      if (action->isParamSetByUser("add_variables") && action->getParam<bool>("add_variables"))
+      {
+        // this sub-action wants the disp-variables added.
+        do_add_variables = true;
+        const auto v = action->getParam<std::vector<VariableName>>("displacements");
+        if (v.size() > 0)
+        {
+          if (displacement_variables.size() == 0)
+            displacement_variables.insert(displacement_variables.end(), v.begin(), v.end());
+          else if (displacement_variables != v)
+            paramError("displacements",
+                       "The vector of displacement variables of the actions differ.");
+        }
+
+        // with block-restriction?
+        if (add_variables_block_restricted && action->isParamSetByUser("block"))
+        {
+          auto action_blocks = action->getParam<std::vector<SubdomainName>>("block");
+          add_variables_blocks.insert(action_blocks.cbegin(), action_blocks.cend());
+        }
+        else
+          add_variables_block_restricted = false;
+
+        // scaling?
+        if (action->getParam<bool>("add_variables") && action->isParamSetByUser("scaling"))
+        {
+          const auto scaling = action->getParam<Real>("scaling");
+
+          // sanity-check
+          if (scaling_set && scaling_value != scaling)
+            paramError("scaling", "The scaling parameter of the actions differ.");
+
+          // set local scaling variables
+          scaling_set = true;
+          scaling_value = scaling;
+        }
+      }
+    }
+
+    // add the disp-variables, if desired
+    if (do_add_variables)
+    {
+
+      auto params = _factory.getValidParams("MooseVariable");
+      // determine necessary order
+      const bool second = _problem->mesh().hasSecondOrderElements();
+
+      params.set<MooseEnum>("order") = second ? "SECOND" : "FIRST";
+      params.set<MooseEnum>("family") = "LAGRANGE";
+      if (add_variables_block_restricted && add_variables_blocks.size() > 0)
+      {
+        std::vector<SubdomainName> blks(add_variables_blocks.begin(), add_variables_blocks.end());
+        params.set<std::vector<SubdomainName>>("block") = blks;
+      }
+      if (scaling_set)
+        params.set<std::vector<Real>>("scaling") = {scaling_value};
+
+      // Loop through the displacement variables
+      if (displacement_variables.size() == 0)
+        paramError("displacements", "The vector of displacement variables is not set.");
```

Comment:
Did we also check that the number of displacement variables == mesh dimension?

Reply:
We should!